### PR TITLE
fix: False positive for StopZoneIdValidator

### PR DESCRIPTION
**Pull request by lionel-nj**
**Summary:**

This PR provides support to avoid false positives in the context of `StopZoneIdValidator`.

**Expected behavior:** 

The validator doesn't trigger `StopWithoutZoneIdNotice` when the fares depend on the routes.

1. Run the fix proposal on [10-15 Transit's GTFS Schedule source](https://transitfeeds.com/p/10-15-transit/936/latest/download). This file has the following structure that is considered valid:

| fare_id | route_id | origin_id | destination_id | contains_id |
|---------|----------|-----------|----------------|-------------|
| 2331    | 7468     |           |                |             |
| 2331    | 7469     |           |                |             |
| 2331    | 7925     |           |                |             |

(from https://github.com/MobilityData/gtfs-validator/issues/1075#issue-1065995820).

2. Observe no more false positive for `StopWithoutZoneIdNotice` in the validation report:

```json
{
  "notices": [
    {
      "code": "stop_without_stop_time",
      "severity": "WARNING",
      "totalNotices": 3,
      "sampleNotices": [
        {
          "csvRowNumber": 85,
          "stopId": "2436381",
          "stopName": "11th Ave & F St"
        },
        {
          "csvRowNumber": 133,
          "stopId": "2729800",
          "stopName": "Vine and Davis"
        },
        {
          "csvRowNumber": 134,
          "stopId": "2729801",
          "stopName": "Vine and Weller"
        }
      ]
    },
    {
      "code": "unknown_column",
      "severity": "INFO",
      "totalNotices": 29,
      "sampleNotices": [
        {
          "filename": "calendar.txt",
          "fieldName": "service_name",
          "index": 2
        },
        {
          "filename": "calendar_dates.txt",
          "fieldName": "holiday_name",
          "index": 3
        },
        {
          "filename": "feed_info.txt",
          "fieldName": "feed_license",
          "index": 5
        },
        {
          "filename": "feed_info.txt",
          "fieldName": "feed_id",
          "index": 10
        },
        {
          "filename": "routes.txt",
          "fieldName": "min_headway_minutes",
          "index": 11
        },
        {
          "filename": "routes.txt",
          "fieldName": "eligibility_restricted",
          "index": 12
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "start_service_area_id",
          "index": 11
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "end_service_area_id",
          "index": 12
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "start_service_area_radius",
          "index": 13
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "end_service_area_radius",
          "index": 14
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "pickup_booking_rule_id",
          "index": 17
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "drop_off_booking_rule_id",
          "index": 18
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "min_arrival_time",
          "index": 19
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "max_departure_time",
          "index": 20
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "mean_duration_factor",
          "index": 21
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "mean_duration_offset",
          "index": 22
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "safe_duration_factor",
          "index": 23
        },
        {
          "filename": "stop_times.txt",
          "fieldName": "safe_duration_offset",
          "index": 24
        },
        {
          "filename": "stops.txt",
          "fieldName": "position",
          "index": 13
        },
        {
          "filename": "stops.txt",
          "fieldName": "direction",
          "index": 14
        },
        {
          "filename": "trips.txt",
          "fieldName": "trip_type",
          "index": 11
        },
        {
          "filename": "trips.txt",
          "fieldName": "drt_max_travel_time",
          "index": 12
        },
        {
          "filename": "trips.txt",
          "fieldName": "drt_avg_travel_time",
          "index": 13
        },
        {
          "filename": "trips.txt",
          "fieldName": "drt_advance_book_min",
          "index": 14
        },
        {
          "filename": "trips.txt",
          "fieldName": "drt_pickup_message",
          "index": 15
        },
        {
          "filename": "trips.txt",
          "fieldName": "drt_drop_off_message",
          "index": 16
        },
        {
          "filename": "trips.txt",
          "fieldName": "continuous_pickup_message",
          "index": 17
        },
        {
          "filename": "trips.txt",
          "fieldName": "continuous_drop_off_message",
          "index": 18
        },
        {
          "filename": "trips.txt",
          "fieldName": "booking_rule_id",
          "index": 19
        }
      ]
    },
    {
      "code": "unknown_file",
      "severity": "INFO",
      "totalNotices": 13,
      "sampleNotices": [
        {
          "filename": "areas.txt"
        },
        {
          "filename": "calendar_attributes.txt"
        },
        {
          "filename": "directions.txt"
        },
        {
          "filename": "fare_rider_categories.txt"
        },
        {
          "filename": "farezone_attributes.txt"
        },
        {
          "filename": "rider_categories.txt"
        },
        {
          "filename": "runcut.txt"
        },
        {
          "filename": "stop_attributes.txt"
        },
        {
          "filename": "timetable_stop_order.txt"
        },
        {
          "filename": "timetables.txt"
        },
        {
          "filename": "linked_datasets.txt"
        },
        {
          "filename": "location_groups.txt"
        },
        {
          "filename": "booking_rules.txt"
        }
      ]
    }
  ]
}
```

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by isabelle-dr** [comment]
Hi @lionel-nj, LGTM!

I tested it on a dataset containing true positives to make sure they are still flagged and it worked. Thanks for updating the documentation 🙏

The true positives were tested on Bustangs's GTFS Schedule dataset ([download link](https://storage.googleapis.com/storage/v1/b/bustang-gtfs-q8996_archives_2021-07-08/o/a24628301fce63ac43d63920c76a5500f5bee3a7.zip?alt=media)) that has 39 instances of this issue. See validation report section below. 

```json
{"code":"stop_without_zone_id",
"severity":"ERROR",
"totalNotices":39,
```

**Review by isabelle-dr — APPROVED** [review]

**Comment by isabelle-dr** [comment]
@lionel-nj Would it be possible to run the acceptance test on this PR to see if any dataset from the Mobility Database would become invalid?
It shouldn't trigger any additional error, since we're fixing false positives here. 

Let's also have a look at how many datasets stop triggering this error after this fix 🎉

**Comment by lionel-nj** [comment]
> @lionel-nj Would it be possible to run the acceptance test on this PR to see if any dataset from the Mobility Database would become invalid?
It shouldn't trigger any additional error, since we're fixing false positives here.

Before the fix, 607/1246 datasets were valid. After the fix, 728/1246 are valid

>Let's also have a look at how many datasets stop triggering this error after this fix 🎉

Before the fix, 263/1247 datasets used to trigger this notice. After the fix, only 74/1246 datasets trigger it. 

In conclusion the number of false positive has decreased and no additional error type is introduced.

**Comment by isabelle-dr** [comment]
Ok, well done 🚀

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopZoneIdValidatorTest.java [hunk 1]
```diff
@@ -71,18 +81,25 @@ private static String toFareRuleId(long csvRowNumber) {
...
-  public void stop_zoneIdNotProvided_yieldsNotice() {
+  public void stop_zoneIdNotProvided_zoneStruncture_yieldsNotice() {
     ImmutableList<GtfsStop> stops = ImmutableList.of(createStop(3, GtfsLocationType.STOP, null));
-    assertThat(generateNotices(stops, ImmutableList.of(createFareRule(5))))
+    assertThat(generateNotices(stops, ImmutableList.of(createFareRuleWithZoneStructure(5))))
         .containsExactly(new StopWithoutZoneIdNotice(stops.get(0)));
   }
 
+  @Test
+  public void stop_zoneIdNotProvided_nozZneStruncture_yieldsNotice() {
```
Comment (asvechnikov2) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopZoneIdValidatorTest.java:91]: nit: s/nozZneStruncture/noZoneStructure/ the same for a test above

File: main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopZoneIdValidatorTest.java [hunk 2]
```diff
@@ -45,10 +45,20 @@ private static GtfsStop createStop(
...
         .setFareId(toFareRuleId(csvRowNumber))
+        .setOriginId("origin id")
+        .build();
+  }
+
+  private static GtfsFareRule createFareRuleWithoutZoneStructure(long csvRowNumber) {
+    return new GtfsFareRule.Builder()
+        .setCsvRowNumber(csvRowNumber)
+        .setFareId(toFareRuleId(csvRowNumber))
+        .setFareId(toFareRuleId(csvRowNumber))
```
Comment (asvechnikov2) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/validator/StopZoneIdValidatorTest.java:60]: nit: duplicated line

**Review by asvechnikov2 — APPROVED** [review]
LGTM!

Do you need to change `StopZoneIdValidator` class docstring?

**Pull request merged**